src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,259 @@
+    // AlterPartitionReassignmentsRequest schemata. Instead of checking for
+    // a null replicas vector in every call to validate_replicas, simply put
+    // those elements first and use iterator magic to run validation on
+    // defined replicas only.
+    auto valid_partitions_end = std::partition(
+      begin, end, [](const reassignable_partition& partition) {
+          return !partition.replicas.has_value();
+      });
+
+    std::vector<reassignable_partition_response> invalid_partitions;
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Empty replica list specified in partition reassignment.",
+      [](const reassignable_partition& partition) {
+          return !partition.replicas->empty();
+      });
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Duplicate replica ids in partition reassignment replica list",
+      [](const reassignable_partition& partition) {
+          auto replicas_copy = *partition.replicas;
+          std::sort(replicas_copy.begin(), replicas_copy.end());
+          auto no_duplicates_it = std::adjacent_find(
+            replicas_copy.cbegin(), replicas_copy.cend());
+          return no_duplicates_it == replicas_copy.cend();
+      });
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Invalid broker id in replica list",
+      [](const reassignable_partition& partition) {
+          auto negative_node_id_it = std::find_if(
+            partition.replicas->begin(),
+            partition.replicas->end(),
+            [](const model::node_id& node_id) { return node_id < 0; });
+          return negative_node_id_it == partition.replicas->end();
+      });
+
+    valid_partitions_end = validate_replicas(
+      valid_partitions_end,
+      end,
+      std::back_inserter(invalid_partitions),
+      error_code::invalid_replica_assignment,
+      "Replica assignment has brokers that are not alive",
+      [&all_node_ids](const reassignable_partition& partition) {
+          auto unkown_broker_id_it = std::find_if(
+            partition.replicas->begin(),
+            partition.replicas->end(),
+            [all_node_ids](const model::node_id& node_id) {
+                return std::find(
+                         all_node_ids.begin(), all_node_ids.end(), node_id)
+                       == all_node_ids.end();
+            });
+          return unkown_broker_id_it == partition.replicas->end();
+      });
+
+    // Store any invalid partitions in the response
+    topic_response.partitions = std::move(invalid_partitions);
+    resp_it = std::move(topic_response);
```

Comment:
nit: this will probably work from what i can tell, but std::transform normally dereferences the output operator. so to be more generic, you might consider `*resp_it = ...`. but this isn't terribly important.

Reply:
std::transform does not touch `resp_it` in this function. Instead, `validate_partitions` adds any invalid partition to the `invalid_parttions` vector and then that vector is put into `resp_it`.

Reply:
> std::transform does not touch resp_it in this function

what i'm saying is that you are effectively writing your own version of a std::transform. if you look at the implementation of std::transform in the standard library you will see that the input which is typically std::back_inserter is dereferenced before assignment. you have exactly the same scenario, but do not dereference the iterator prior to assignment.

Reply:
Gotcha. I made the change.